processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -331,90 +334,147 @@ public static void validateRowValues(
     final Indexed<String> dim1Names = adapter1.getDimensionNames();
     final Indexed<String> dim2Names = adapter2.getDimensionNames();
     for (int i = 0; i < dims1.length; ++i) {
-      final int[] dim1Vals = dims1[i];
-      final int[] dim2Vals = dims2[i];
+      final Object dim1Vals = dims1[i];
```

Comment:
ah, I think I see. there are checks for dictionary encoding all through here. That doesn't seem like a sustainable solution. Is it possible to have the checks be done in a way that allows the capabilities to control the type assumptions?